reasoner/planner/RecursivePlanner.java

```diff
@@ -0,0 +1,382 @@
+            Set<Variable> resolvableMode = Collections.intersection(estimateableVariables(resolvable), currentBounds);
+            initialiseDependencies(conjunctionNode, resolvable, resolvableMode);
+            if (!resolvable.isNegated()) currentBounds.addAll(estimateableVariables(resolvable));
+        }
+    }
+
+    private void initialiseDependencies(ConjunctionNode conjunctionNode, Resolvable<?> resolvable, Set<Variable> resolvableMode) {
+        if (resolvable.isConcludable()) {
+            Set<ResolvableConjunction> cyclicDependencies = conjunctionNode.cyclicDependencies(resolvable.asConcludable());
+            for (CallMode callMode : triggeredCalls(resolvable.asConcludable(), resolvableMode, Optional.empty())) {
+                recursivelyGeneratePlanChoices(callMode);
+                if (!cyclicDependencies.contains(callMode.conjunction)) {
+                    plan(callMode); // Acyclic dependencies can be fully planned
+                }
+            }
+        } else if (resolvable.isNegated()) {
+            iterate(resolvable.asNegated().disjunction().conjunctions()).forEachRemaining(conjunction -> {
+                CallMode callMode = new CallMode(conjunction, resolvableMode);
+                recursivelyGeneratePlanChoices(callMode);
+                plan(callMode);
+            });
+        }
+    }
+
+    private void recursivelyGeneratePlanChoices(CallMode callMode) {
+        if (!planChoices.containsKey(callMode)) {
+            planChoices.put(callMode, null); // Guard
+            ConjunctionNode conjunctionNode = conjunctionGraph.conjunctionNode(callMode.conjunction);
+            answerCountEstimator.buildConjunctionModel(callMode.conjunction);
+
+            Map<Set<Pair<Concludable, Set<Variable>>>, PlanChoice> planChoices = new HashMap<>();
+            PartialOrderReductionSearch porSearch = new PartialOrderReductionSearch(logicMgr.compile(callMode.conjunction), callMode.mode);
+            for (List<Resolvable<?>> ordering : porSearch.allOrderings()) {
+                initialiseOrderingDependencies(conjunctionNode, ordering, callMode.mode);
+                PlanChoice planChoice = summarisePlanChoice(conjunctionNode, ordering, callMode.mode);
+
+                // Two orderings for the same CallMode with identical cyclic-concludable modes are interchangeable in the subgraph-plan
+                //      -> We only need to keep the cheaper one.
+                if (!planChoices.containsKey(planChoice.cyclicConcludableModes) ||
+                        planChoices.get(planChoice.cyclicConcludableModes).acyclicCost > planChoice.acyclicCost) {
+                    planChoices.put(planChoice.cyclicConcludableModes, planChoice);
+                }
+            }
+            this.planChoices.put(callMode, new HashSet<>(planChoices.values()));
+        }
+    }
+
+    private PlanChoice summarisePlanChoice(ConjunctionNode conjunctionNode, List<Resolvable<?>> ordering, Set<Variable> mode) {
```

Comment:
do we still want to keep the term "summarise?

Reply:
No, but it's a bit hard to think of a better term. createPlanChoice?

Reply:
better, make sure you change it to createOrderingChoice if you go with that naming